src/Components/Components/src/PersistentStateValueProvider.cs

```diff
@@ -52,17 +69,34 @@ public void Subscribe(ComponentState subscriber, in CascadingParameterInfo param
     {
         var propertyName = parameterInfo.PropertyName;
         var propertyType = parameterInfo.PropertyType;
-        _subscriptions[subscriber] = state.RegisterOnPersisting(() =>
+        _subscriptions[subscriber] = state.RegisterOnPersisting(async () =>
             {
                 var storageKey = ComputeKey(subscriber, propertyName);
                 var propertyGetter = ResolvePropertyGetter(subscriber.Component.GetType(), propertyName);
                 var property = propertyGetter.GetValue(subscriber.Component);
                 if (property == null)
                 {
-                    return Task.CompletedTask;
+                    return;
                 }
+
+                // Try to get a custom serializer for this type first
+                var serializerType = typeof(IPersistentComponentStateSerializer<>).MakeGenericType(propertyType);
+                var customSerializer = serviceProvider.GetService(serializerType);
+                
+                if (customSerializer != null)
+                {
+                    // Use reflection to call the generic PersistAsync method with the custom serializer
+                    var persistMethod = typeof(PersistentComponentState).GetMethod(nameof(PersistentComponentState.PersistAsync), BindingFlags.Instance | BindingFlags.Public, [typeof(string), propertyType, serializerType, typeof(CancellationToken)]);
+                    if (persistMethod != null)
+                    {
+                        var task = (Task)persistMethod.Invoke(state, [storageKey, property, customSerializer, CancellationToken.None])!;
+                        await task;
+                        return;
+                    }
+                }
```

Comment:
Cache the serializers by type on a `ConcurrentDictionary<Type, IPersistentComponentStateSerializer>` (see how it's done in this file.

IPersistentComponentStateSerializer should be an internal interface, have the same methods as IPersistentComponentStateSerializer but take the `Type type` argument as a first argument.

`IPersistentComponentStateSerializer<T>` should implement `IPersistentComponentStateSerializer` with a default interface implementation that calls into Persist<T> no matter the Type t passed as first argument

Reply:
Move this code outside of the lambda.